File: libraries/imageeditor-native/src/main/cpp/spoiler/StackBlur.cpp
// Copyright (c) 2026 Element Creations Ltd.
// SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial.
#include "StackBlur.h"

#include <algorithm>
#include <cstdlib>
#include <vector>

namespace photoedit::spoiler {

void stackBlur(uint32_t* pix, int w, int h, int radius) {
    if (radius < 1 || w <= 0 || h <= 0) return;

    const int wm = w - 1;
    const int hm = h - 1;
    const int div = radius + radius + 1;
    const int r1 = radius + 1;
    const int wh = w * h;

    // Per-pixel intermediate buffers — the horizontal pass writes to r/g/b, the
    // vertical pass reads from them. Allocated once per call; for the spoiler use
    // case the bitmap is 20 dp wide (~60 px on a 3x display) so wh is in the
    // hundreds, well below the cost of any heap noise.
    std::vector<int> r(wh);
    std::vector<int> g(wh);
    std::vector<int> b(wh);
    std::vector<int> vmin(std::max(w, h));

    int divsum = (div + 1) >> 1;
    divsum *= divsum;
    std::vector<int> dv(256 * divsum);
    for (size_t i = 0; i < dv.size(); ++i) {
        dv[i] = static_cast<int>(i) / divsum;
    }

    // Circular stack of (r, g, b) triples, holding the running window of pixels in
    // the current row/column. Length is `div = 2 * radius + 1`.
    std::vector<int> stackR(div), stackG(div), stackB(div);

    // ----- Horizontal pass -----
    int yw = 0;
    int yi = 0;
    for (int y = 0; y < h; ++y) {
        int rinsum = 0, ginsum = 0, binsum = 0;
        int routsum = 0, goutsum = 0, boutsum = 0;
        int rsum = 0, gsum = 0, bsum = 0;

        for (int i = -radius; i <= radius; ++i) {
            const uint32_t p = pix[yi + std::min(wm, std::max(i, 0))];
            const int si = i + radius;
            stackR[si] = static_cast<int>((p >> 16) & 0xff);
            stackG[si] = static_cast<int>((p >> 8) & 0xff);
            stackB[si] = static_cast<int>(p & 0xff);
            const int rbs = r1 - std::abs(i);
            rsum += stackR[si] * rbs;
            gsum += stackG[si] * rbs;
            bsum += stackB[si] * rbs;
            if (i > 0) {
                rinsum += stackR[si]; ginsum += stackG[si]; binsum += stackB[si];
            } else {
                routsum += stackR[si]; goutsum += stackG[si]; boutsum += stackB[si];
            }
        }
        int stackpointer = radius;

        for (int x = 0; x < w; ++x) {
            r[yi] = dv[rsum];
            g[yi] = dv[gsum];
            b[yi] = dv[bsum];
            rsum -= routsum; gsum -= goutsum; bsum -= boutsum;
            const int stackstart = (stackpointer - radius + div) % div;
            routsum -= stackR[stackstart]; goutsum -= stackG[stackstart]; boutsum -= stackB[stackstart];

            if (y == 0) vmin[x] = std::min(x + radius + 1, wm);
            const uint32_t p = pix[yw + vmin[x]];
            stackR[stackstart] = static_cast<int>((p >> 16) & 0xff);
            stackG[stackstart] = static_cast<int>((p >> 8) & 0xff);
            stackB[stackstart] = static_cast<int>(p & 0xff);
            rinsum += stackR[stackstart]; ginsum += stackG[stackstart]; binsum += stackB[stackstart];
            rsum += rinsum; gsum += ginsum; bsum += binsum;

            stackpointer = (stackpointer + 1) % div;
            routsum += stackR[stackpointer]; goutsum += stackG[stackpointer]; boutsum += stackB[stackpointer];
            rinsum -= stackR[stackpointer]; ginsum -= stackG[stackpointer]; binsum -= stackB[stackpointer];
            ++yi;
        }
        yw += w;
    }

    // ----- Vertical pass -----
    for (int x = 0; x < w; ++x) {
        int rinsum = 0, ginsum = 0, binsum = 0;
        int routsum = 0, goutsum = 0, boutsum = 0;
        int rsum = 0, gsum = 0, bsum = 0;
        int yp = -radius * w;

        for (int i = -radius; i <= radius; ++i) {
            yi = std::max(0, yp) + x;
            const int si = i + radius;
            stackR[si] = r[yi]; stackG[si] = g[yi]; stackB[si] = b[yi];
            const int rbs = r1 - std::abs(i);
            rsum += r[yi] * rbs;
            gsum += g[yi] * rbs;
            bsum += b[yi] * rbs;
            if (i > 0) {
                rinsum += stackR[si]; ginsum += stackG[si]; binsum += stackB[si];
            } else {
                routsum += stackR[si]; goutsum += stackG[si]; boutsum += stackB[si];
            }
            if (i < hm) yp += w;
        }

        yi = x;
        int stackpointer = radius;
        for (int y = 0; y < h; ++y) {
            pix[yi] = (0xffu << 24)
                | (static_cast<uint32_t>(dv[rsum]) << 16)
                | (static_cast<uint32_t>(dv[gsum]) << 8)
                | static_cast<uint32_t>(dv[bsum]);
            rsum -= routsum; gsum -= goutsum; bsum -= boutsum;
            const int stackstart = (stackpointer - radius + div) % div;
            routsum -= stackR[stackstart]; goutsum -= stackG[stackstart]; boutsum -= stackB[stackstart];

            if (x == 0) vmin[y] = std::min(y + r1, hm) * w;
            const int p = x + vmin[y];
            stackR[stackstart] = r[p]; stackG[stackstart] = g[p]; stackB[stackstart] = b[p];
            rinsum += stackR[stackstart]; ginsum += stackG[stackstart]; binsum += stackB[stackstart];
            rsum += rinsum; gsum += ginsum; bsum += binsum;

            stackpointer = (stackpointer + 1) % div;
            routsum += stackR[stackpointer]; goutsum += stackG[stackpointer]; boutsum += stackB[stackpointer];
            rinsum -= stackR[stackpointer]; ginsum -= stackG[stackpointer]; binsum -= stackB[stackpointer];
            yi += w;
        }
    }
}

} // namespace photoedit::spoiler

File: libraries/imageeditor-native/src/main/kotlin/io/element/android/libraries/imageeditor/native_/ui/components/StackBlur.kt
/*
 * Copyright (c) 2026 Element Creations Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial.
 * Please see LICENSE files in the repository root for full details.
 */
package io.element.android.libraries.imageeditor.native_.ui.components

import android.graphics.Bitmap
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * In-place stack blur for [Bitmap]s. Ported from Mario Klingemann's original
 * `StackBlur` (CC-BY 3.0) — the same algorithm Telegram-Android calls from its native
 * `image.cpp:stackBlurBitmap`. Two-pass integer arithmetic, runs in microseconds on the
 * small 20-dp-wide spoiler backdrop bitmaps we feed it.
 *
 * Why we ship our own port instead of `Modifier.blur`: Compose's `Modifier.blur` is a
 * RenderNode shader applied at composition time — the kernel reads from the underlying
 * layer buffer, which means the parent's background colour bleeds INTO the photo at
 * partial-alpha edges. On a black-background photo the bleed shows up as grey/white,
 * exactly the symptom the user reported. Operating on the bitmap bytes before display
 * sidesteps that entirely: the dst bitmap is opaque end-to-end, and no compositing
 * surface is involved.
 *
 * Alpha is forced to 0xFF on every output pixel. The spoiler backdrop is never meant to
 * be transparent, and writing solid alpha here removes any chance of edge-bleed into the
 * surrounding background regardless of the source bitmap's alpha channel.
 */
internal object StackBlur {
    fun blur(bitmap: Bitmap, radius: Int) {
        if (radius < 1) return
        // Native path first — see cpp/spoiler/StackBlur.cpp. JNI dispatch + locked
        // pixel access skips the IntArray copy this Kotlin fallback has to do, so on
        // hardware where libphotoedit.so loaded the cost is ~10× lower.
        if (SpoilerJni.available) {
            SpoilerJni.stackBlur(bitmap, radius)
            return
        }
        val w = bitmap.width
        val h = bitmap.height
        if (w <= 0 || h <= 0) return
        val pix = IntArray(w * h)
        bitmap.getPixels(pix, 0, w, 0, 0, w, h)

        val wm = w - 1
        val hm = h - 1
        val div = radius + radius + 1
        val r = IntArray(w * h)
        val g = IntArray(w * h)
        val b = IntArray(w * h)
        val vmin = IntArray(max(w, h))

        var divsum = (div + 1) shr 1
        divsum *= divsum
        val dv = IntArray(256 * divsum)
        for (i in dv.indices) dv[i] = i / divsum

        var yw = 0
        var yi = 0
        val stack = Array(div) { IntArray(3) }
        val r1 = radius + 1

        // ----- Horizontal pass -----
        for (y in 0 until h) {
            var rinsum = 0; var ginsum = 0; var binsum = 0
            var routsum = 0; var goutsum = 0; var boutsum = 0
            var rsum = 0; var gsum = 0; var bsum = 0
            for (i in -radius..radius) {
                val p = pix[yi + min(wm, max(i, 0))]
                val sir = stack[i + radius]
                sir[0] = (p shr 16) and 0xff
                sir[1] = (p shr 8) and 0xff
                sir[2] = p and 0xff
                val rbs = r1 - abs(i)
                rsum += sir[0] * rbs
                gsum += sir[1] * rbs
                bsum += sir[2] * rbs
                if (i > 0) {
                    rinsum += sir[0]; ginsum += sir[1]; binsum += sir[2]
                } else {
                    routsum += sir[0]; goutsum += sir[1]; boutsum += sir[2]
                }
            }
            var stackpointer = radius

            for (x in 0 until w) {
                r[yi] = dv[rsum]
                g[yi] = dv[gsum]
                b[yi] = dv[bsum]
                rsum -= routsum; gsum -= goutsum; bsum -= boutsum
                val stackstart = stackpointer - radius + div
                var sir = stack[stackstart % div]
                routsum -= sir[0]; goutsum -= sir[1]; boutsum -= sir[2]

                if (y == 0) vmin[x] = min(x + radius + 1, wm)
                val p = pix[yw + vmin[x]]
                sir[0] = (p shr 16) and 0xff
                sir[1] = (p shr 8) and 0xff
                sir[2] = p and 0xff
                rinsum += sir[0]; ginsum += sir[1]; binsum += sir[2]
                rsum += rinsum; gsum += ginsum; bsum += binsum

                stackpointer = (stackpointer + 1) % div
                sir = stack[stackpointer]
                routsum += sir[0]; goutsum += sir[1]; boutsum += sir[2]
                rinsum -= sir[0]; ginsum -= sir[1]; binsum -= sir[2]
                yi++
            }
            yw += w
        }

        // ----- Vertical pass -----
        for (x in 0 until w) {
            var rinsum = 0; var ginsum = 0; var binsum = 0
            var routsum = 0; var goutsum = 0; var boutsum = 0
            var rsum = 0; var gsum = 0; var bsum = 0
            var yp = -radius * w
            for (i in -radius..radius) {
                yi = max(0, yp) + x
                val sir = stack[i + radius]
                sir[0] = r[yi]; sir[1] = g[yi]; sir[2] = b[yi]
                val rbs = r1 - abs(i)
                rsum += r[yi] * rbs
                gsum += g[yi] * rbs
                bsum += b[yi] * rbs
                if (i > 0) {
                    rinsum += sir[0]; ginsum += sir[1]; binsum += sir[2]
                } else {
                    routsum += sir[0]; goutsum += sir[1]; boutsum += sir[2]
                }
                if (i < hm) yp += w
            }
            yi = x
            var stackpointer = radius
            for (y in 0 until h) {
                pix[yi] = (0xff shl 24) or (dv[rsum] shl 16) or (dv[gsum] shl 8) or dv[bsum]
                rsum -= routsum; gsum -= goutsum; bsum -= boutsum
                val stackstart = stackpointer - radius + div
                var sir = stack[stackstart % div]
                routsum -= sir[0]; goutsum -= sir[1]; boutsum -= sir[2]

                if (x == 0) vmin[y] = min(y + r1, hm) * w
                val p = x + vmin[y]
                sir[0] = r[p]; sir[1] = g[p]; sir[2] = b[p]
                rinsum += sir[0]; ginsum += sir[1]; binsum += sir[2]
                rsum += rinsum; gsum += ginsum; bsum += binsum

                stackpointer = (stackpointer + 1) % div
                sir = stack[stackpointer]
                routsum += sir[0]; goutsum += sir[1]; boutsum += sir[2]
                rinsum -= sir[0]; ginsum -= sir[1]; binsum -= sir[2]
                yi += w
            }
        }

        bitmap.setPixels(pix, 0, w, 0, 0, w, h)
    }
}
